src/dm/DispatchManagerActions.cc

```diff
@@ -2249,7 +2305,760 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+{
+    ostringstream oss;
+
+    bool is_update = false;
+
+    vector<VectorAttribute *> sg_rules;
+
+    // -------------------------------------------------------------------------
+    // Check action consistency:
+    // - Object exists VM, NIC and SG
+    // - State is compatible with operation
+    // -------------------------------------------------------------------------
+    if ( auto vm = vmpool->get(vid) )
+    {
+        VirtualMachine::LcmState lstate = vm->get_lcm_state();
+
+        switch (lstate)
+        {
+            //Cannnot update VM, SG rules being updated/created
+            case VirtualMachine::BOOT:
+            case VirtualMachine::BOOT_MIGRATE:
+            case VirtualMachine::BOOT_SUSPENDED:
+            case VirtualMachine::BOOT_STOPPED:
+            case VirtualMachine::BOOT_UNDEPLOY:
+            case VirtualMachine::BOOT_POWEROFF:
+            case VirtualMachine::BOOT_UNKNOWN:
+            case VirtualMachine::BOOT_FAILURE:
+            case VirtualMachine::BOOT_MIGRATE_FAILURE:
+            case VirtualMachine::BOOT_UNDEPLOY_FAILURE:
+            case VirtualMachine::BOOT_STOPPED_FAILURE:
+            case VirtualMachine::MIGRATE:
+            case VirtualMachine::HOTPLUG_NIC:
+            case VirtualMachine::HOTPLUG_NIC_POWEROFF:
+            case VirtualMachine::UNKNOWN:
+                oss << "VM " << vid << " is in wrong state " << vm->state_str();
+                error_str = oss.str();
+
+                return -1;
+
+            //Update SG rules at host
+            case VirtualMachine::RUNNING:
+            case VirtualMachine::HOTPLUG:
+            case VirtualMachine::HOTPLUG_SNAPSHOT:
+            case VirtualMachine::HOTPLUG_SAVEAS:
+            case VirtualMachine::HOTPLUG_RESIZE:
+            case VirtualMachine::DISK_SNAPSHOT:
+            case VirtualMachine::DISK_SNAPSHOT_DELETE:
+            case VirtualMachine::DISK_RESIZE:
+                is_update = true;
+                break;
+
+            default:
+                break;
+        }
+
+        auto nic = vm->get_nic(nicid);
+
+        if ( nic == nullptr )
+        {
+            oss << "VM " << vid << " doesn't have NIC id " << nicid;
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        set<int> sgs;
+        nic->get_security_groups(sgs);
+
+        if (sgs.find(sgid) == sgs.end())
+        {
+            oss << "VM " << vid << " NIC " << nicid << " doesn't contain SG " << sgid;
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        // ----------------------------------------------------------------------
+        // Update VM data
+        // - NIC to remove SG
+        // - Template with remaining SG rules (could be empty)
+        // ----------------------------------------------------------------------
+        nic->remove_security_group(sgid);
+
+        vm->remove_security_group(sgid);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

User input is being used directly in C++ file system operations without proper validation. This can allow attackers to access files outside the intended directory through path traversal attacks, potentially exposing sensitive information or enabling system compromise. Use `fs::canonical()` to resolve paths and validate they start with an allowed base directory, implement allowlist validation using `std::find()` to check against permitted directories, or sanitize input by removing dangerous sequences like &#34;..&#34; using string operations. [Learn more](https://cwe.mitre.org/data/definitions/22.html)